executor/oci/spec.go

```diff
@@ -11,3 +31,233 @@ const (
 	// NoProcessSandbox should be enabled only when the BuildKit is running in a container as an unprivileged user.
 	NoProcessSandbox
 )
+
+// Ideally we don't have to import whole containerd just for the default spec
+
+// GenerateSpec generates spec using containerd functionality.
+// opts are ignored for s.Process, s.Hostname, and s.Mounts .
+func GenerateSpec(ctx context.Context, meta executor.Meta, mounts []executor.Mount, id, resolvConf, hostsFile string, namespace network.Namespace, processMode ProcessMode, idmap *idtools.IdentityMapping, opts ...oci.SpecOpts) (*specs.Spec, func(), error) {
+	c := &containers.Container{
+		ID: id,
+	}
+
+	// containerd/oci.GenerateSpec requires a namespace, which
+	// will be used to namespace specs.Linux.CgroupsPath if generated
+	if _, ok := namespaces.Namespace(ctx); !ok {
+		ctx = namespaces.WithNamespace(ctx, "buildkit")
+	}
+
+	if securityOpts, err := generateSecurityOpts(meta.SecurityMode); err == nil {
+		opts = append(opts, securityOpts...)
+	} else {
+		return nil, nil, err
+	}
+
+	if processModeOpts, err := generateProcessModeOpts(processMode); err == nil {
+		opts = append(opts, processModeOpts...)
+	} else {
+		return nil, nil, err
+	}
+
+	s, err := oci.GenerateSpec(ctx, nil, c, opts...)
+	if err != nil {
+		return nil, nil, err
+	}
+	// set the networking information on the spec
+	namespace.Set(s)
+
+	s.Process.Args = meta.Args
+	s.Process.Env = meta.Env
+	s.Process.Cwd = meta.Cwd
+	s.Process.Rlimits = nil           // reset open files limit
+	s.Process.NoNewPrivileges = false // reset nonewprivileges
+	s.Hostname = "buildkitsandbox"
+
+	// Setup for a Linux-based container (includes LCOW)
+	if s.Linux != nil {
+		s.Mounts, err = GetMounts(ctx,
+			withProcessMode(processMode),
+			withROBind(resolvConf, "/etc/resolv.conf"),
+			withROBind(hostsFile, "/etc/hosts"),
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		s.Mounts = append(s.Mounts, specs.Mount{
+			Destination: "/sys/fs/cgroup",
+			Type:        "cgroup",
+			Source:      "cgroup",
+			Options:     []string{"ro", "nosuid", "noexec", "nodev"},
+		})
+
+		if processMode == NoProcessSandbox {
+			var maskedPaths []string
+			for _, s := range s.Linux.MaskedPaths {
```

Comment:
how is this not host specific if it sets up `Linux` struct of the spec?

Reply:
If the spec has both `Linux !=nil` and `Windows != nil`, then it'll be run in LCOW mode. Because we call [`oci.GenerateSpec`](https://pkg.go.dev/github.com/containerd/containerd/oci?tab=doc#GenerateSpec), we'll never get `Linux != nil` on Windows. If and when we support LCOW, then we need to call [`oci.GenerateSpecWithPlatform`](https://pkg.go.dev/github.com/containerd/containerd/oci?tab=doc#GenerateSpecWithPlatform) instead. and if we specify Linux platform on Windows, we'll get `Linux != nil`, and can then fill-out the relevant parts.

Whether this particular code is appropriate for LCOW, I don't know. I haven't tried LCOW support, so I don't actually know how the filesystem mounts work. It's possible `NoProcessSandbox` is invalid/nonsensical for LCOW anyway, but working that out seems like a less-interesting excusion, given that containerd doesn't have the filesystem mounting/diffing support set up for LCOW anyway.

Reply:
Let's not do anything for LCOW atm. It is abandoned in moby as well and code is getting removed when there is a refactor. cc @thaJeztah

Reply:
Okay. I can move this block out into another helper function in spec_unix.go, with an empty function in spec_window.go.

If I'm going to do that, I might-as-well add a check in spec_windows.go (which I used locally in testing) which will fail the generation if it finds a non-nil `s.Linux`.

I'm sorry to hear LCOW is being dropped by moby. I was rather looking forward to seeing how LCOW would evolve to the WSL2 (actual Linux kernel) approach. I assume the refactor you're talking about will instead deliver support for a single client talking to a WCOW dockerd and a dockerd inside WSL2?